File: misc/code/top-k-frequency-trie.java
import java.util.HashMap;
import java.util.Map;
import java.util.PriorityQueue;

class TrieNode {
    Map<Character, TrieNode> children;
    int count;

    public TrieNode() {
        children = new HashMap<>();
        count = 0;
    }
}

class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String name) {
        TrieNode node = root;
        for (char ch : name.toCharArray()) {
            node = node.children.computeIfAbsent(ch, k -> new TrieNode());
        }
        node.count++;
    }

    public void getAllNames(TrieNode node, StringBuilder prefix, PriorityQueue<NameCount> minHeap, int k) {
        if (node == null) return;
        if (node.count > 0) {
            if (minHeap.size() < k) {
                minHeap.offer(new NameCount(prefix.toString(), node.count));
            } else if (node.count > minHeap.peek().count) {
                minHeap.poll();
                minHeap.offer(new NameCount(prefix.toString(), node.count));
            }
        }
        for (Map.Entry<Character, TrieNode> entry : node.children.entrySet()) {
            prefix.append(entry.getKey());
            getAllNames(entry.getValue(), prefix, minHeap, k);
            prefix.deleteCharAt(prefix.length() - 1);
        }
    }

    public PriorityQueue<NameCount> getTopKNames(int k) {
        PriorityQueue<NameCount> minHeap = new PriorityQueue<>(k);
        getAllNames(root, new StringBuilder(), minHeap, k);
        return minHeap;
    }
}

class NameCount implements Comparable<NameCount> {
    String name;
    int count;

    public NameCount(String name, int count) {
        this.name = name;
        this.count = count;
    }

    @Override
    public int compareTo(NameCount other) {
        return Integer.compare(this.count, other.count);
    }

    @Override
    public String toString() {
        return name + ": " + count;
    }
}

public class Main {
    public static void main(String[] args) {
        String[] names = {"张伟", "王伟伟", "王芳", "李伟", "李娜"}; // 示例数据
        int k = 100; // 找到前100个重名人数最多的姓名

        Trie trie = new Trie();
        for (String name : names) {
            trie.insert(name);
        }

        PriorityQueue<NameCount> topKNames = trie.getTopKNames(k);
        while (!topKNames.isEmpty()) {
            System.out.println(topKNames.poll());
        }
    }
}

File: misc/code/top-k-frequency-trie.cpp
#include <unordered_map>
#include <string>
#include <vector>
#include <queue>
#include <algorithm> // min()
using namespace std;

struct trie_node {
    unordered_map<char, trie_node *> children;
    int count;

public:
    trie_node() : count(0) {}
};

struct name_count {
    string name;
    int count;

    bool operator<(const name_count &other) const {
        return this->count > other.count;
    }
};



class Trie {
private:
    trie_node *root;

public:
    Trie() {
        root = new trie_node();
    }

    void insert(string name) {
        trie_node *p = root;
        for (char c : name) {
            p = p->children[c] == nullptr ? (p->children[c] = new trie_node()) : p->children[c];
        }
        p->count++;
    }

    // dfs 遍历每个节点，并随时更新 topk-heap。minHeap 和 k 两个参数，也可以作为类成员变量。
    void traverse_trie(trie_node *node, string &prefix, priority_queue<name_count> &minHeap, int k) {
        if (node == nullptr) {
            return;
        }
        // 处理：是否放入 topk
        if (node->count > 0) {
            if (minHeap.size() < k) {
                minHeap.push({prefix, node->count});
            } else if (node->count > minHeap.top().count) {
                minHeap.pop();
                minHeap.push({prefix, node->count});
            }
        }
        // dfs 遍历
        for (auto &i : node->children) {
            prefix.push_back(i.first);
            traverse_trie(i.second, prefix, minHeap, k);
            prefix.pop_back();
        }
    }

    vector<name_count> get_topk_names(int k) {
        priority_queue<name_count> minHeap;
        string prefix = "";
        traverse_trie(root, prefix, minHeap, k);
        k = min(k, (int) minHeap.size());
        vector<name_count> res(k);
        while (!minHeap.empty()) {
            res[--k] = minHeap.top();
            minHeap.pop();
        }
        return res;
    }
};

int main() {
    string names[] = {"张伟", "王伟伟", "张伟", "李那", "李那边", "王芳", 
            "王伟伟", "张伟", "李那", "李那边", "王芳", "李伟", "张伟", "李那",
            "李那边", "李这", "李这边", "张伟光正"}; // 示例数据
    int k = 100; // 找到前100个重名人数最多的姓名

    Trie t;
    for (string name : names) {
        t.insert(name);
    }
    vector<name_count> topk_names = t.get_topk_names(k);
    for (const name_count &nc : topk_names) {
        printf("name:%s, count:%d\n", nc.name.c_str(), nc.count);
    }
    return 0;
}
